compiler/rustc_mir/src/transform/lower_slice_len.rs

```diff
@@ -0,0 +1,144 @@
+//! This pass lowers calls to core::slice::len to just Len op.
+//! It should run before inlining!
+
+use crate::transform::MirPass;
+use rustc_hir::def_id::DefId;
+use rustc_middle::mir::visit::MutVisitor;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, TyCtxt};
+
+pub struct LowerSliceLenCalls;
+
+impl<'tcx> MirPass<'tcx> for LowerSliceLenCalls {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        lower_slice_len_calls(tcx, body)
+    }
+}
+
+pub fn lower_slice_len_calls<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+    // FIXME: (shamatar) Can be tuned later
+    // if tcx.sess.mir_opt_level() < 3 {
+    //     return;
+    // }
+
+    // Optimization is once per basic block, so we just visit terminators
+    let languate_items = tcx.lang_items();
+    let slice_lang_item_def_id = if let Some(slice_lang_item) = languate_items.slice_impl() {
+        slice_lang_item
+    } else {
+        // there is no language item to compare too :)
+        return;
+    };
+    let len_symbol = rustc_span::Symbol::intern("len");
+
+    let mut visitor = LowerLenCallVisitor { tcx, len_symbol, slice_lang_item_def_id };
+    visitor.visit_body(body);
```

Comment:
I don't see much benefit of a visitor over manually iterating over the basic blocka in this case.

Reply:
There are quite a few short circuit early returns in the implementation, so moving it to the different circuit and only visiting basic blocks reduced brackets depth

Reply:
I mean you can use `for block in body.basic_blocks_mut()` instead of a full blown visitor.

Reply:
yes, this is how we do intrinsic lowering in https://github.com/rust-lang/rust/blob/5bc1f917838ca8af6fce4ee9fed19f5cb534f989/compiler/rustc_mir/src/transform/lower_intrinsics.rs#L15-L18

maybe we should rename that pass to lower_call and split it up so there's one path for intrinsics and one path for lang items?

Reply:
I can restructure in this form. A question remains how to deal with Symbol interning to avoid calls for it at every single body

Reply:
I like bjorn3's idea of just making the len function a lang item and not needing any kind of symbol lookup

Reply:
It's partially reworked in the last commit, but I do not understand how to make a new language item